spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SaferObjectInputStream.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.common.util;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.NotSerializableException;
+import java.io.ObjectInputStream;
+import java.io.ObjectStreamClass;
+import java.util.Collections;
+import java.util.List;
+import org.springframework.util.ClassUtils;
+
+/**
+ * Special ObjectInputStream subclass that checks if classes are allowed to deserialize.
+ * The class should be configured with a whitelist of only allowed (safe) classes to deserialize.
+ *
+ * @author Artem Smotrakov
+ */
+public class SaferObjectInputStream extends ObjectInputStream {
```

Comment:
Let's rename this to `WhitelistedObjectInputStream`.

Can we also make this a `private class` internal (inner) to `SerializationStrategy`.

Reply:
Sure, no problem. However, it can't be a private class since `SerializationStrategy` is now an interface.

Reply:
I'll restrict the access to package-private.